### PR TITLE
refactor(store-core): migrate git operation handlers (#3125)

### LINKED ISSUE #3125: refactor(store-core): migrate git operation handlers (#2661)
**Parent:** #2661

Migrate git-operation message handlers (callback-style) into `@chroxy/store-core/handlers`. Same shape as #3124 — extract payload normalization, leave callback wiring per-platform.

## Cases to migrate

- `diff_result` — `{files: DiffFile[], error}`
- `git_status_result` — `{branch, staged: GitFileStatus[], unstaged: GitFileStatus[], untracked: string[], error}`
- `git_branches_result` (app-only) — `{branches: GitBranch[], currentBranch, error}`
- `git_stage_result` / `git_unstage_result` (app-only, share same handler) — `{error}`
- `git_commit_result` (app-only) — payload TBD by reading the existing app handler

## Approach

1. Add to `packages/store-core/src/handlers/index.ts`:
   - `handleDiffResult(msg)`
   - `handleGitStatusResult(msg)`
   - `handleGitBranchesResult(msg)`
   - `handleGitStageResult(msg)` (shared by stage and unstage)
   - `handleGitCommitResult(msg)`

2. Each returns a normalized payload type. Keep entry types (`DiffFile`, `GitFileStatus`, `GitBranch`) as `unknown[]` placeholders — concrete types live downstream.

3. Wire dashboard + app message-handlers to import the shared extractors. Dashboard currently only has `diff_result` and `git_status_result` cases — preserve that scope (no new cases on dashboard).

4. Add tests covering each handler: valid payload, missing fields default to null/[], wrong-type fields default to []/null.

## Acceptance criteria

- [ ] All 5 handlers exported from `@chroxy/store-core`
- [ ] Dashboard message-handler uses shared extractors for `diff_result` and `git_status_result`
- [ ] App message-handler uses shared extractors for all 5 cases
- [ ] Tests cover happy path + missing/invalid field defaults
- [ ] No behaviour change
- [ ] Server tests + lint + dashboard tests + app type-check all pass

### PR DESCRIPTION
**Pull request by blamechris**
## Summary

Migrates git-operation message handlers into `@chroxy/store-core` so the dashboard and app stop duplicating payload parsing.

Closes #3125. Refs #2661 (parent).

## What's shared

| Handler | Shape | Returns |
|---|---|---|
| `handleDiffResult` | Callback payload | `{files: unknown[], error: string \| null}` |
| `handleGitStatusResult` | Callback payload | `{branch, staged, unstaged, untracked, error}` |
| `handleGitBranchesResult` | Callback payload (app-only) | `{branches, currentBranch, error}` |
| `handleGitStageResult` | Callback payload (app-only, shared by stage+unstage) | `{error}` |
| `handleGitCommitResult` | Callback payload (app-only) | `{hash, message, error}` |

Concrete element types (`DiffFile`, `GitFileStatus`, `GitBranch`) stay downstream in each consumer — the shared handlers keep entries as `unknown[]` to avoid pulling consumer types up into store-core. Per-element shape is NOT validated; matches the prior inline `as DiffFile[]` casts.

Behaviour preserved exactly:
- `branch`, `currentBranch`, `error`, `hash`, `message` use bare `typeof === 'string'` (no trim, empty strings preserved verbatim) — matches the inline guards both clients used.
- `files`, `staged`, `unstaged`, `untracked`, `branches` fall back to `[]` when missing or non-array.
- Dashboard wires only `diff_result` and `git_status_result` (preserves prior scope — no new cases on dashboard).
- App wires all five message types; `git_stage_result` and `git_unstage_result` share the same handler since their payloads are identical.

Callback dispatch (`get()._diffCallback`, `getCallback('gitStatus')`, etc.) stays platform-specific — only payload normalization is shared.

## Tests

28 new vitest cases covering: valid payloads, missing fields → null/[] fallback, wrong-type fields → defaults. Empty-string preservation tested for fields that previously used bare `typeof` checks.

## Verification

- `cd packages/store-core && npm test` — 364 passed
- `cd packages/dashboard && npm run typecheck && npm test` — typecheck clean, 1290 passed
- `cd packages/app && npx tsc --noEmit && npm test` — typecheck clean, 1142 passed

## Test plan

- [ ] CI green (server tests + lint + app type check + dashboard type check)
- [ ] Spot-check dashboard `diff` panel still populates after invoking `_diffCallback`
- [ ] Spot-check dashboard `git status` panel still populates with branch/staged/unstaged/untracked
- [ ] Spot-check app `git_branches_result`, `git_stage_result`, `git_unstage_result`, `git_commit_result` still drive their respective callbacks

### CODE REVIEW

**Comment by blamechris** [comment]
## Code Review: PR #3128

### Summary

Migrates 5 git-operation message handlers (`diff_result`, `git_status_result`, `git_branches_result`, `git_stage_result`/`git_unstage_result`, `git_commit_result`) from inline duplication in `packages/app/src/store/message-handler.ts` and `packages/dashboard/src/store/message-handler.ts` into shared functions in `@chroxy/store-core`. Closes #3125, refs #2661 (parent meta-issue).

The migration is mechanical and behaviour-preserving — same `Array.isArray(...) ? ... as T[] : []` and bare `typeof === 'string' ? ... : null` guards as before, just relocated. Callback dispatch (`get()._diffCallback` vs `getCallback('diff')`) stays platform-specific, only payload normalisation is shared.

### Strengths

- **Pattern consistency.** The new handlers follow the exact same convention as the recently merged sibling PR #3124 (file operation handlers): `unknown[]` placeholders for entries, downstream concrete-type casts, and behaviour-preserving inline guards. This makes the cumulative migration into `store-core` audit-friendly.
- **Faithful behaviour preservation.** The PR description explicitly calls out that empty strings pass through verbatim (using bare `typeof` rather than `parseStringField`) — and the tests assert that explicitly (`'preserves empty branch string verbatim'`, `'preserves empty hash/message strings verbatim'`). This is the right call for a mechanical migration.
- **Strong test coverage.** 28 new vitest cases covering valid payloads, missing fields, wrong-type fields, and the empty-string preservation contract. Handler bodies are trivial; the tests are the real safeguard against accidental drift.
- **Comprehensive doc comments.** The 14-line section header at line 1488 explains exactly why entry types stay `unknown[]`, why bare `typeof` is preferred over `parseStringField`, and notes that tightening would be a behaviour change out of scope. Future maintainers will not have to reverse-engineer this decision.
- **Stage/unstage de-duplication.** Recognising that `git_stage_result` and `git_unstage_result` have identical payloads and routing both cases through `handleGitStageResult` is a nice DRY win that the inline code did not capture.

### Issues Found

#### Critical (Must Fix)

None.

#### Suggestions (Should Consider)

| File | Line | Suggestion | Rationale |
|------|------|------------|-----------|
| `packages/store-core/src/handlers/index.ts` | 1488 (block header) | Add a one-liner deferring stricter element validation to a future PR | The header notes that per-element validation is out of scope, but no follow-up issue tracks it. If the team ever wants real `DiffFile`/`GitFileStatus`/`GitBranch` validation, that intent will be lost in the comment. Tracked as deferred item below. |
| `packages/app/src/store/message-handler.ts` | 2127–2135, 2139–2150, 2154–2162 | Consider `as unknown as DiffFile[]` instead of bare `as DiffFile[]` | The shared handler returns `unknown[]`, and casting `unknown[]` directly to `DiffFile[]` works but TypeScript tooling (e.g. lint rules in stricter projects) sometimes flags this. The bare cast matches the existing style in the codebase, so this is purely stylistic. |
| `packages/store-core/src/handlers/index.ts` | 1492 (header) | Header comments mention "five share the callback-style shape" but `git_stage_result`/`git_unstage_result` are technically two cases sharing one handler — minor wording nit | The "five handlers" framing is fine since each output type is distinct, but a future reader counting `case` statements might see six message types. A brief parenthetical "(stage and unstage share one)" would remove ambiguity. |

#### Nitpicks (Optional)

- The new doc-comment paragraphs in `handlers/index.ts` are heavy on prose. The same style is used by sibling handlers (`handleDirectoryListing`, etc.) so consistency wins, but if any future polish pass tightens those, this batch would benefit too.
- `handleGitStageResult`'s body is identical to other "error-only" handlers (e.g. `handleWriteFileResult` minus the path). A future micro-refactor could introduce a shared `parseErrorOnlyPayload(msg)` helper, but it would only save 3 lines per call site — not worth the indirection today.

### Deferred Items (Follow-Up Issues)

| Suggestion | Issue | Rationale for deferral |
|------------|-------|------------------------|
| Add per-element validation for `DiffFile`, `GitFileStatus`, `GitBranch` payloads | [#3132](https://github.com/blamechris/chroxy/issues/3132) | The PR's explicit scope is mechanical migration with zero behaviour change; tightening element shapes is a separate concern that needs its own design (Zod schema, fail-soft vs fail-loud, telemetry on rejection). |

### Architecture Notes

This PR is one tile in the long-running #2661 migration: collapse `app/src/store/message-handler.ts` (~2,200 lines) and `dashboard/src/store/message-handler.ts` (~2,200 lines) into a shared `@chroxy/store-core` parser layer. Each prior tile (auth+connection #3108, conversation+history #3118, multi-client #3111, error+log #3110, checkpoints #3109, permission flow #3117, session lifecycle #3120, file ops #3124) followed the same pattern, so consumers and tooling are well-rehearsed. After this merges, the remaining git-related cases on the dashboard (`git_branches_result`, `git_stage_result`, `git_unstage_result`, `git_commit_result`) can adopt the shared handlers in a follow-up without further store-core changes — exactly the goal.

The decision to keep `unknown[]` in the shared handler and cast at the call site is the right one — pulling `DiffFile`/`GitFileStatus`/`GitBranch` types up into `store-core` would couple it to dashboard/app types and undermine the platform-agnostic contract.

### CI

All required checks were passing at review time (Server Tests, Server Lint, App Tests, App Type Check, Dashboard Type Check, Store Core Tests, Store Core Type Check, Protocol Tests, App Expo Doctor); Desktop Rust Tests and Dashboard Tests were still pending (unrelated to this change set).

### Verdict

- [x] **Approve** — Ready to merge

Mechanical migration done well, follows the established pattern, behaviour-preserving by construction, and the test coverage matches sibling PRs. No critical issues.

File: packages/store-core/src/handlers/handlers.test.ts [hunk 1]
```diff
@@ -2018,3 +2023,211 @@ describe('handleSessionSwitched', () => {
...
+    expect(
+      handleGitCommitResult({ error: 'nothing to commit' }).error,
+    ).toBe('nothing to commit')
+  })
+
+  it('preserves empty hash/message strings verbatim', () => {
+    // Inline impl uses `typeof === 'string' ? ... : null`, so empty strings pass through.
+    const result = handleGitCommitResult({ hash: '', message: '' })
+    expect(result.hash).toBe('')
+    expect(result.message).toBe('')
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/store-core/src/handlers/handlers.test.ts]: **FIX**

Fixed in `fa2159d18`

**Change:** Extended the existing 'preserves empty hash/message strings verbatim' test to also cover `error: ''`, since the inline impl uses bare `typeof === 'string'` for that field too. Renamed for accuracy.

```diff
- it('preserves empty hash/message strings verbatim', () => {
-   // Inline impl uses `typeof === 'string' ? ... : null`, so empty strings pass through.
-   const result = handleGitCommitResult({ hash: '', message: '' })
-   expect(result.hash).toBe('')
-   expect(result.message).toBe('')
- })
+ it('preserves empty hash/message/error strings verbatim', () => {
+   // Inline impl uses `typeof === 'string' ? ... : null`, so empty strings pass through.
+   const result = handleGitCommitResult({ hash: '', message: '', error: '' })
+   expect(result.hash).toBe('')
+   expect(result.message).toBe('')
+   expect(result.error).toBe('')
+ })
```

File: packages/store-core/src/handlers/index.ts [hunk 2]
```diff
@@ -1485,3 +1485,156 @@ export function handlePermissionRulesUpdated(
...
+// each consumer — the shared handlers keep entries as `unknown[]` to avoid
+// pulling concrete types up into store-core. Per-element shape is NOT
+// validated here; matches the inline `as DiffFile[]` casts both clients used
+// prior to this migration. Tightening would be a behaviour change and is out
+// of scope for the #2661 mechanical migration.
+// ---------------------------------------------------------------------------
+
+/** Parsed payload from a `diff_result` message. */
+export interface DiffResultPayload {
+  /** Files entries — element shape validated downstream. */
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/store-core/src/handlers/index.ts]: **FIX**

Fixed in `fa2159d18`

**Change:** Reworded the `files` field doc on `DiffResultPayload` to make explicit that elements are passed through verbatim (no runtime validation). Matches the section header above the type, and prevents readers from assuming validation that isn't performed.

```diff
- /** Files entries — element shape validated downstream. */
+ /** File entries passed through verbatim; elements are left as `unknown[]` and cast/handled by consumers. */
  files: unknown[]
```

File: packages/store-core/src/handlers/index.ts [hunk 3]
```diff
@@ -1485,3 +1485,156 @@ export function handlePermissionRulesUpdated(
...
+/** Parsed payload from a `git_status_result` message. */
+export interface GitStatusResultPayload {
+  /** Current branch name, or null when missing/non-string. */
+  branch: string | null
+  /** Staged file entries — element shape validated downstream. */
+  staged: unknown[]
+  /** Unstaged file entries — element shape validated downstream. */
+  unstaged: unknown[]
+  /** Untracked file paths (validated as array; element shape validated downstream). */
+  untracked: unknown[]
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/store-core/src/handlers/index.ts:1777]: **FIX**

Fixed in `fa2159d18`

**Change:** Reworded the `staged`, `unstaged`, and `untracked` field docs on `GitStatusResultPayload` to make explicit that elements are passed through verbatim. Same wording as the `files` doc above.

```diff
- /** Staged file entries — element shape validated downstream. */
+ /** Staged file entries passed through verbatim; elements are left as `unknown[]` and cast/handled by consumers. */
  staged: unknown[]
- /** Unstaged file entries — element shape validated downstream. */
+ /** Unstaged file entries passed through verbatim; elements are left as `unknown[]` and cast/handled by consumers. */
  unstaged: unknown[]
- /** Untracked file paths (validated as array; element shape validated downstream). */
+ /** Untracked file paths (validated as array); elements are left as `unknown[]` and cast/handled by consumers. */
  untracked: unknown[]
```

File: packages/store-core/src/handlers/index.ts [hunk 4]
```diff
@@ -1485,3 +1485,156 @@ export function handlePermissionRulesUpdated(
...
+    staged: Array.isArray(msg.staged) ? (msg.staged as unknown[]) : [],
+    unstaged: Array.isArray(msg.unstaged) ? (msg.unstaged as unknown[]) : [],
+    untracked: Array.isArray(msg.untracked) ? (msg.untracked as unknown[]) : [],
+    error: typeof msg.error === 'string' ? msg.error : null,
+  }
+}
+
+/** Parsed payload from a `git_branches_result` message (app-only today). */
+export interface GitBranchesResultPayload {
+  /** Branch entries — element shape validated downstream. */
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/store-core/src/handlers/index.ts]: **FIX**

Fixed in `fa2159d18`

**Change:** Reworded the `branches` field doc on `GitBranchesResultPayload` to make explicit that elements are passed through verbatim, matching the `files`/`staged`/`unstaged` rewords above.

```diff
- /** Branch entries — element shape validated downstream. */
+ /** Branch entries passed through verbatim; elements are left as `unknown[]` and cast/handled by consumers. */
  branches: unknown[]
```

File: packages/store-core/src/handlers/handlers.test.ts [hunk 5]
```diff
@@ -2018,3 +2023,211 @@ describe('handleSessionSwitched', () => {
...
+
+  it('returns null error when missing or non-string', () => {
+    expect(handleGitStatusResult({}).error).toBeNull()
+    expect(handleGitStatusResult({ error: 99 }).error).toBeNull()
+  })
+
+  it('preserves empty branch string verbatim', () => {
+    // Inline impl uses `typeof === 'string' ? ... : null`, so empty strings pass through.
+    expect(handleGitStatusResult({ branch: '' }).branch).toBe('')
+  })
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/store-core/src/handlers/handlers.test.ts:2284]: **FIX**

Fixed in `fa2159d18`

**Change:** Added a 'preserves empty error string verbatim' test for `handleGitStatusResult`, mirroring the existing empty-branch test. Locks in the bare `typeof === 'string'` semantics for the `error` field.

```diff
  it('preserves empty branch string verbatim', () => {
    // Inline impl uses `typeof === 'string' ? ... : null`, so empty strings pass through.
    expect(handleGitStatusResult({ branch: '' }).branch).toBe('')
  })
+
+ it('preserves empty error string verbatim', () => {
+   // Inline impl uses `typeof === 'string' ? ... : null`, so empty strings pass through.
+   expect(handleGitStatusResult({ error: '' }).error).toBe('')
+ })
```

File: packages/store-core/src/handlers/handlers.test.ts [hunk 6]
```diff
@@ -2018,3 +2023,211 @@ describe('handleSessionSwitched', () => {
...
+
+  it('returns empty array when branches is missing or non-array', () => {
+    expect(handleGitBranchesResult({}).branches).toEqual([])
+    expect(handleGitBranchesResult({ branches: 'x' }).branches).toEqual([])
+    expect(handleGitBranchesResult({ branches: null }).branches).toEqual([])
+  })
+
+  it('returns null currentBranch when missing or non-string', () => {
+    expect(handleGitBranchesResult({}).currentBranch).toBeNull()
+    expect(handleGitBranchesResult({ currentBranch: 42 }).currentBranch).toBeNull()
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/store-core/src/handlers/handlers.test.ts]: **FIX**

Fixed in `fa2159d18`

**Change:** Added 'preserves empty currentBranch and error strings verbatim' test for `handleGitBranchesResult`. Covers both bare-`typeof` string fields in one test, locking in the empty-string passthrough semantics.

```diff
  it('forwards branch elements verbatim — no per-element validation', () => {
    const branches = [{ junk: 'x' }, null, 7]
    expect(handleGitBranchesResult({ branches }).branches).toBe(branches)
  })
+
+ it('preserves empty currentBranch and error strings verbatim', () => {
+   // Inline impl uses `typeof === 'string' ? ... : null`, so empty strings pass through.
+   expect(handleGitBranchesResult({ currentBranch: '' }).currentBranch).toBe('')
+   expect(handleGitBranchesResult({ error: '' }).error).toBe('')
+ })
```

**Comment by blamechris** [comment]
## Review Comments Addressed

| # | Comment | Outcome | Reference |
|---|---------|---------|-----------|
| 1 | `handleGitCommitResult` empty-string test should also cover `error: ''` | FIX | `fa2159d18` |
| 2 | `DiffResultPayload.files` doc implies validation that doesn't happen | FIX | `fa2159d18` |
| 3 | `GitStatusResultPayload.staged/unstaged/untracked` docs imply validation that doesn't happen | FIX | `fa2159d18` |
| 4 | `GitBranchesResultPayload.branches` doc implies validation that doesn't happen | FIX | `fa2159d18` |
| 5 | `handleGitStatusResult` missing empty-string test for `error` | FIX | `fa2159d18` |
| 6 | `handleGitBranchesResult` missing empty-string test for `currentBranch` | FIX | `fa2159d18` |

**Total:** 6 comments addressed
- Fixed: 6 (commit `fa2159d18`)
- False positives: 0
- Follow-up issues created: 0
- Existing issues closed: 0

All fixes are doc-only and test-only — zero behaviour change. Locally verified: `npm test` passes 366 tests (was 364, +2 net), `npm run typecheck` clean.
